servers/src/main/java/tachyon/master/file/FileSystemMaster.java

```diff
@@ -1340,33 +1415,35 @@ boolean unmountInternal(TachyonURI tachyonPath) throws InvalidPathException {
     return mMountTable.delete(tachyonPath);
   }
 
-
   /**
    * MasterInodeTTL periodic check.
    */
-  public final class MasterInodeTTLCheckExecutor implements HeartbeatExecutor {
-    // TODO: current implementation needs to be improved by using a more efficient datastructure
-    // such as hourly bucketized inodefilelist
+  private final class MasterInodeTTLCheckExecutor implements HeartbeatExecutor {
     @Override
     public void heartbeat()  {
       synchronized (mInodeTree) {
-        List<Inode> inodes = mInodeTree.getInodeChildrenRecursive(mInodeTree.getRoot());
-        for (Inode inode : inodes) {
-          if (inode.isFile() && !inode.isPinned()) {
-            InodeFile iFile = (InodeFile) inode;
-            long ttl = iFile.getTTL();
-            if (ttl > 0 && System.currentTimeMillis() - iFile.getCreationTimeMs() > ttl * 1000) {
+        if (mTTLBuckets.size() > 0 && mTTLBuckets.get(0).getTTLIntervalStartTimeMs() < 0) {
+          // The ttl interval start of this bucket is smaller than 0, so all files in this bucket
+          // should no longer live.
+          for (InodeFile file : mTTLBuckets.get(0).getFiles()) {
+            if (!file.isDeleted() && !file.isPinned()) {
               try {
-                deleteFile(iFile.getId(), false);
+                deleteFile(file.getId(), false);
               } catch (FileDoesNotExistException e) {
-                LOG.error("file does not exit " + iFile.toString());
+                LOG.error("file does not exit " + file.toString());
               } catch (InvalidPathException e) {
-                LOG.error("invalid path for ttl check " + iFile.toString());
+                LOG.error("invalid path for ttl check " + file.toString());
               } catch (IOException e) {
-                LOG.error("IO exception for ttl check" + iFile.toString());
+                LOG.error("IO exception for ttl check" + file.toString());
               }
             }
           }
+          mTTLBuckets.remove(0);
+        }
+
+        // Subtract elapsed time from start time of all buckets.
+        for (TTLBucket bucket : mTTLBuckets) {
+          bucket.setTTLIntervalStartMs(bucket.getTTLIntervalStartTimeMs() - mTTLCheckerIntervalMs);
```

Comment:
Actually after thinking it twice, I think here is the tradeoff that you either need to do more work in finding the bucket index on insert (on create path), or spend more time to update each bucket (on heartbeat path). It seems create path is more mission-critical and you are probably making the right decision. But you may want to document it.

Also, using SkipList to organize buckets can compromise this tradeoff :) maybe leave it  as a TODO.